client/rpc-servers/src/middleware.rs

```diff
@@ -93,149 +97,96 @@ impl RpcMetrics {
-	type CallFuture = FutureOutput;
-
-	fn on_request<F, X>(
-		&self,
-		request: jsonrpc_core::Request,
-		meta: M,
-		next: F,
-	) -> Either<Self::Future, X>
-	where
-		F: Fn(jsonrpc_core::Request, M) -> X + Send + Sync,
-		X: Future<Output = Option<jsonrpc_core::Response>> + Send + 'static,
-	{
-		let metrics = self.metrics.clone();
-		let transport_label = self.transport_label.clone();
-		if let Some(ref metrics) = metrics {
-			metrics.requests_started.with_label_values(&[transport_label.as_str()]).inc();
-		}
-		let r = next(request, meta);
-		Either::Left(
-			async move {
-				let r = r.await;
-				if let Some(ref metrics) = metrics {
-					metrics.requests_finished.with_label_values(&[transport_label.as_str()]).inc();
-				}
-				r
-			}
-			.boxed(),
-		)
+impl Middleware for RpcMiddleware {
+	type Instant = std::time::Instant;
+
+	fn on_connect(&self) {
+		self.metrics.ws_sessions_opened.as_ref().map(|counter| counter.inc());
 	}
 
-	fn on_call<F, X>(
-		&self,
-		call: jsonrpc_core::Call,
-		meta: M,
-		next: F,
-	) -> Either<Self::CallFuture, X>
-	where
-		F: Fn(jsonrpc_core::Call, M) -> X + Send + Sync,
-		X: Future<Output = Option<jsonrpc_core::Output>> + Send + 'static,
-	{
-		let start = std::time::Instant::now();
-		let name = call_name(&call, &self.known_rpc_method_names).to_owned();
-		let metrics = self.metrics.clone();
-		let transport_label = self.transport_label.clone();
-		log::trace!(target: "rpc_metrics", "[{}] {} call: {:?}", transport_label, name, &call);
-		if let Some(ref metrics) = metrics {
-			metrics
-				.calls_started
-				.with_label_values(&[transport_label.as_str(), name.as_str()])
-				.inc();
-		}
-		let r = next(call, meta);
-		Either::Left(
-			async move {
-				let r = r.await;
-				let micros = start.elapsed().as_micros();
-				if let Some(ref metrics) = metrics {
-					metrics
-						.calls_time
-						.with_label_values(&[transport_label.as_str(), name.as_str()])
-						.observe(micros as _);
-					metrics
-						.calls_finished
-						.with_label_values(&[
-							transport_label.as_str(),
-							name.as_str(),
-							if is_success(&r) { "true" } else { "false" },
-						])
-						.inc();
-				}
-				log::debug!(
-					target: "rpc_metrics",
-					"[{}] {} call took {} μs",
-					transport_label,
-					name,
-					micros,
-				);
-				r
-			}
-			.boxed(),
-		)
+	fn on_request(&self) -> Self::Instant {
+		let now = std::time::Instant::now();
+		self.metrics.requests_started.with_label_values(&[self.transport_label]).inc();
+		now
 	}
-}
 
-fn call_name<'a>(call: &'a jsonrpc_core::Call, known_methods: &HashSet<String>) -> &'a str {
-	// To prevent bloating metric with all invalid method names we filter them out here.
-	let only_known = |method: &'a String| {
-		if known_methods.contains(method) {
-			method.as_str()
-		} else {
-			"invalid method"
-		}
-	};
+	fn on_call(&self, name: &str) {
+		log::trace!(target: "rpc_metrics", "[{}] on_call name={}", self.transport_label, name);
```

Comment:
This format is also breaking but not possible to get which kind a method was called here.

In the future it would be quite neat the know the params too here e.g. the number of keys for `state_storageSubscription`

Reply:
For this PR we should stick as closely to `jsonrpc` as possible. Not sure if I understood correctly: does `jsonrpc` include the params and we're not, and we're not because we can't (atm) because of `Middleware` limitations?

Will current metrics break completely because of this or is it just going to display less information?

Reply:
> For this PR we should stick as closely to jsonrpc as possible. Not sure if I understood correctly: does jsonrpc include the params and we're not, and we're not because we can't (atm) because of Middleware limitations?

Yes, `jsonrpc` prints the entire call for example `[ws] system_health call: MethodCall(MethodCall { jsonrpc: Some(V2), method: "system_health", params: Array([]), id: Num(34) })`

jsonrpsee prints: `[ws] on_call name=state_subscribeStorage`

I have no idea how these "metrics logs" are used but the prometheus "should" be the important stuff and it's the same.
That's what you can find in grafana for example

/cc @gabreal @PierreBesson do you know? :point_up:

Reply:
ah, so it's "breaking" because it doesn't **log** exactly the same, but Prometheus is not affected, I get it now.

I think this is fine. Not great, but ok. Let's log an issue to investigate if we can improve `Middleware` to have access to the params, but I hope our metrics do not depend on specific log strings.